DRILL-6340: Output Batch Control in Project using the RecordBatchSizer

bitblender opened the pull request:
Changes required to implement Output Batch Sizing in Project using the RecordBatchSizer.

ilooner commented on common/src/test/resources/logback-test.xml:
```diff
@@ -50,7 +50,7 @@
 
   <root>
-    <level value="error"/>
+    <level value="error"/>o
```
typo

ilooner commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java:
```diff
@@ -193,6 +209,13 @@ protected IterOutcome doWork() {
           incomingRecordCount = incoming.getRecordCount();
+          memoryManager.update();
+          if (logger.isTraceEnabled()) {
+            logger.trace("doWork():[1] memMgr RC " + memoryManager.getOutputRowCount()
```
use {} instead of concatenating strings with +

bitblender replied:
Ok. Will do. But what is the advantage of using '{}' over static string concatenation with a check? I think format strings are more error(argument ordering errors) prone than concatenation.

ilooner replied:
It's easier to read for the programmer. Also the string concatenation is only executed if the debug statement is executed. If you use + you have to have if statements around your debug statements to avoid any performance impact.

```
if (logger.isDebug()) {
 logger.debug("my " + variable + " message");
}
```

If you use the {} you don't have to have if statements in order to prevent a performance impact

```
logger.debug("my {} message", variable);
```

It's typically considered best practice to do the latter since it's much cleaner.

ilooner commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java:
```diff
@@ -206,12 +229,24 @@ protected IterOutcome doWork() {
+
+    int maxOuputRecordCount = memoryManager.getOutputRowCount();
+    if (logger.isTraceEnabled()) {
+      logger.trace("doWork():[2] memMgr RC " + memoryManager.getOutputRowCount()
```
Same as above.

ilooner commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java:
```diff
@@ -206,12 +229,24 @@ protected IterOutcome doWork() {
+    final int outputRecords = projector.projectRecords(this.incoming,0, maxOuputRecordCount, 0);
+    long projectEndTime = System.currentTimeMillis();
+    if (logger.isTraceEnabled()) {
+      logger.trace("doWork(): projection: " + " records " + outputRecords + ", time " + (projectEndTime - projectStartTime) + " ms");
```
Use {}

ilooner commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java:
```diff
@@ -230,18 +265,37 @@ protected IterOutcome doWork() {
     }
-    final int projRecords = projector.projectRecords(remainderIndex, remainingRecordCount, 0);
+    if (logger.isTraceEnabled()) {
+      logger.trace("handleRemainder: remaining RC " + remainingRecordCount + " toProcess " + recordsToProcess
```
{}

ilooner commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java:
```diff
@@ -497,30 +561,36 @@ private void setupNewSchemaFromInput(RecordBatch incomingBatch) throws SchemaCha
+
+    long setupNewSchemaEndTime = System.currentTimeMillis();
+    if (logger.isTraceEnabled()) {
+      logger.trace("setupNewSchemaFromInput: time " + (setupNewSchemaEndTime - setupNewSchemaStartTime) + " ms" + ", proj " + this + " incoming " + incomingBatch);
```
{}

bitblender commented:
I don't think my tests are contributing a whole lot to the slowdown. With this change, Project will split large batches in tests(like 45 M batches in TestLargeFileCompilation) to ~16M batches. This somehow seems to be slowing down operators like Sort and TopN a fair bit compared to the no-split runs. This needs further investigation to see if these slow-down are un-avoidable .

Ben-Zvi commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/OutputWidthVisitor.java:
```diff
@@ -0,0 +1,278 @@
+        OutputWidthCalculator widthCalculator = holder.getOutputWidthCalculator();
+        final int argSize = holderExpr.args.size();
+        ArrayList<OutputWidthExpression> arguments = null;
+        if (argSize != 0) {
```
Trivial question: Would it also work without that if() check ?  I.e., use an empty ArrayList in case of zero args.

Ben-Zvi commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectMemoryManager.java:
```diff
@@ -0,0 +1,315 @@
+ *
+ *
+ * Execution phase: Just before a batch is processed by Project, the PMM walks the tree of OutputWidthExpressions
+ * and conversts them to FixedWidthExpressions. It uses the RecordBatchSizer and the function annotations to do this conversion.
```
"converts"

Ben-Zvi commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectMemoryManager.java:
```diff
@@ -0,0 +1,315 @@
+        rowWidth += totalVariableColumnWidth;
+        int outPutRowCount;
+        if (rowWidth != 0) {
+            setOutputRowCount(getOutputBatchSize(), rowWidth);
```
Maybe add a little comment here; this is the important place in the code where the output size for Project is being set ....

ilooner commented:
@bitblender Looks like this is pretty close to being done and has already been +1'ed. Could you just fix conflicts and push your latest changes addressing the review comments, and indicate which comments you addressed and which comments you decided to leave as is. After that I'll add the **ready-to-commit** tag.

bitblender commented:
@ilooner I have addressed all the comments and I am running the changes through the pre-commit framework. You can go ahead and mark it as ready-to-commit

bitblender commented:
@ilooner @Ben-Zvi let me know when you are done looking at my latest round of changes in response to the review comments. I will rebase on master, fix the conflict and force push.

ilooner commented:
Thanks @bitblender +1 LGTM
